general-concepts/bundled-dependencies: new section

thesamesam opened the pull request:
I've tried to faithfully port the wiki page [0] to the devmanual in this commit, and intend to change the contents as required in followups, to allow easier comparison and to retain provenance.

[0] https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies

Closes: https://bugs.gentoo.org/300625

----

~Note: I'm looking for review of the formatting and porting to the devmanual for now, not whether we should add/adjust content etc (which I will do once the foundation is OK).~

thesamesam committed:
general-concepts/bundled-deps: new section

I've tried to faithfully port the wiki page [0] to the devmanual in
this commit, and intend to change the contents as required in followups,
to allow easier comparison and to retain provenance.

[0] https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies

Closes: https://bugs.gentoo.org/300625
Signed-off-by: Sam James <sam@gentoo.org>

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```diff
@@ -0,0 +1,373 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<devbook self="ebuild-writing/bundled-dependencies/">
+<chapter>
+<title>Bundled dependencies</title>
+
+<body>
+<p>
```
"Blank lines must be placed immediately after every `<body>` (opening tag only) and before every `<section>`, `<p>`, `<pre>`, `<codesample>`, `<figure>`, `<table>`, `<ul>`, `<ol>`, `<dl>`, `<note>`, `<important>` and `<warning>` (opening tags only)."
```suggestion
<title>Bundled dependencies</title>
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```diff
@@ -0,0 +1,373 @@
+</p>
+
+<ul>
+  <li>
+    Statically linking against a system library
+  </li>
+  <li>
+   Shipping and using your own copy of a library
+  </li>
+  <li>
+   Including and (unconditionally) using snippets of code copied from
+   a library
+  </li>
```
Consistent indentation please (+ 2 positions per level).

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```diff
@@ -0,0 +1,373 @@
+  </li>
+</ul>
+
+<p>In other words, code bundling occurs whenever a program or library ends
+up containing code that does not belong to it.</p>
```
```suggestion
<p>
In other words, code bundling occurs whenever a program or library ends
up containing code that does not belong to it.
</p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```diff
@@ -0,0 +1,373 @@
+
+<section>
+<title>Temptations</title>
+<body>
+<p>
```
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```diff
@@ -0,0 +1,373 @@
+
+<subsection>
+<title>Private forks</title>
+<body>
+<p>
```
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```diff
@@ -0,0 +1,373 @@
+
+<subsection>
+<title>Security implications</title>
+<body>
+<p>
```
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```diff
@@ -0,0 +1,373 @@
+
+<p>
+Now, depending on how bad things are:
+</p>
+<ul>
```
```suggestion
</p>

<ul>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```diff
@@ -0,0 +1,373 @@
+
+<p>
+Examples:
+</p>
+<ul>
```
```suggestion
</p>

<ul>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```suggestion
</p>

<ul>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/bundled-dependencies/text.xml:
```suggestion
<body>

<p>
```

ulm commented on ebuild-writing/text.xml:
```diff
@@ -31,4 +31,5 @@ with some general notes and extended examples.
 <include href="misc-files/"/>
 <include href="user-submitted/"/>
 <include href="common-mistakes/"/>
+<include href="bundled-dependencies/"/>
```
Can you use a shorter name please, like `bundled-deps`?

(There's of course nothing fundamentally wrong with the current filename, but with the hierarchical structure filenames tend to become very long, and so far we've tried to countermand this by using short path components.)

ulm replied:
Sorry about the 17 pedantic comments, but you've explicitly asked for a "review of the formatting". :smile:

thesamesam replied:
I was in two minds about "dependencies" vs "deps", and I wasn't sure if the abbreviation of "deps" was undesirable somehow. I'll change it.

ulm commented on ebuild-writing/bundled-deps/text.xml:
```diff
@@ -0,0 +1,391 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<devbook self="ebuild-writing/bundled-deps/">
+<chapter>
+<title>Bundled dependencies</title>
```
Yes, of course. We have this in many places, for example:
```
<?xml version="1.0" encoding="UTF-8"?>
<devbook self="quickstart/">
<chapter>
<title>Quickstart ebuild guide</title>
```

thesamesam commented:
How's it looking now? OK to proceed to content review? And do we want to commit this as-is, or review the content here? Either is fine with me. I guess reviewing the content here is easier because you can comment on the full diff more easily.

What I don't want to do, however, is squash any content fixes into the first commit.

ulm commented on ebuild-writing/bundled-deps/text.xml:
```diff
@@ -0,0 +1,390 @@
+  <li>
+    If <e>foo</e> bundled local copy of <e>libbar</e>, then they would have to wait
+    till you discover the vulnerability, update <e>libbar</e> sources, release
+    the new version and distributions package the new version.
```
No full stop here.

ulm commented on ebuild-writing/bundled-deps/text.xml:
Neither here.

ulm commented on ebuild-writing/bundled-deps/text.xml:
What are these stray `#` in the text? Leftover from a numbered list in RestructuredText or Markdown?

thesamesam replied:
Ah, it's numbered list markup from mediawiki...

ulm commented on ebuild-writing/bundled-deps/text.xml:
Update the link for the new filename. And the section part is missing a `#`:
```suggestion
<uri link="::ebuild-writing/bundled-deps/#Downstream consequences">
```

ulm commented on ebuild-writing/bundled-deps/text.xml:
Again, no full stop.

ulm commented on ebuild-writing/bundled-deps/text.xml:
The blank lines before the `<p>` tags are not needed within list items. (Sorry, I had missed this in the first round.)

thesamesam commented:
Let me know when it looks OK and I'll move onto content (I don't want to try fix existing style issues in the first commit once I started that, as cherry-picking that will be hell).

ulm reviewed (commented):
Formatting looks good.

I have some tiny comments, admittedly most are into spelling territory (but you might want to fix them now, so they won't interfere with content review later).

thesamesam commented on ebuild-writing/bundled-deps/text.xml:
```diff
@@ -0,0 +1,394 @@
+<body>
+
+<p>
+Especially in Windows, shipping dependencies <e>can</e> be a favor to users
```
Ah, a subtle wiki->devmanual issue to keep in mind..

ulm reviewed (commented):
Formatting LGTM.

laumann commented on ebuild-writing/bundled-deps/text.xml:
```diff
@@ -0,0 +1,392 @@
+Bundling <e>D</e> hides the dependency on <e>D</e> in a way: if the packager
+is not paying close attention <e>P</e> may even get in despite and with the
+bundled dependency. (It is, however, only a matter of time until someone
+noticed the bundling.)
```
noticed → notices

laumann commented on ebuild-writing/bundled-deps/text.xml:
```diff
@@ -0,0 +1,392 @@
+</p>
+
+<p>
+Now, a very important security flaw has been found in <e>libbar</e>
```
"very" is a weasel word - maybe "critical security flaw"?

laumann reviewed (commented):
This is good reading :+1:

idk if you want examples of packages where upstream does vendor dependencies, but has a mechanism not to use them. media-libs/openjpeg vendors some libraries that Gentoo's packaging carefully removes. At least it's optional to use the vendored versions.

thesamesam committed:
general-concepts/bundled-deps: 3rd-person perspective; grammar tweaks

* Use 3rd-person perspective as we do elsewhere
* Tweak grammar (usually just a missing comma or so)
* Use longer example names where it aids readability (rather than letters;
  kept in some places)

Signed-off-by: Sam James <sam@gentoo.org>

thesamesam committed:
general-concepts/bundled-deps: mention "bundled vs vulnerable"

A common answer when asked to unbundle is that projects will simply
update their bundled copy when a vulnerability is discovered.

This has two issues:
1) nobody checks if old versions of packages are vulnerable;
2) MITRE does not issue CVEs for bundled versions of dependencies.

It's therefore easy to miss that a copy is problematically stale.

Suggested-by: Michael Orlitzky <mjo@gentoo.org>
Signed-off-by: Sam James <sam@gentoo.org>

thesamesam committed:
general-concepts/bundled-deps: mention testsuites, meson subprojects

bug #919100 is a nice example of where Meson subprojects exposed a real
problem in a bundled dependency.

Bug: https://bugs.gentoo.org/919100
Signed-off-by: Sam James <sam@gentoo.org>
